File: 1497-check-if-array-pairs-are-divisible-by-k/1497-check-if-array-pairs-are-divisible-by-k.cpp
class Solution {
public:
    bool canArrange(vector<int>& arr, int k) {
        unordered_map<int,int> m;
        for(int i=0;i<arr.size();i++){
            int rem = ((arr[i]%k)+k)%k; // to handle both -ve and +ve nos.
            m[rem]++;
        }
        if(m[0]%2!=0)
            return false;
        
        for(int i=1;i<=k/2;i++){
            if(m[i] != m[k-i])
                return false;
        }
        
        return true;
    }
};





/*
(a%k+b%k)%k = 0, if (a+b)%k = 0
problem converts to 2 sum problem
using remainders
*/

File: 1497-check-if-array-pairs-are-divisible-by-k/1497-check-if-array-pairs-are-divisible-by-k.java
class Solution {
    public boolean canArrange(int[] arr, int k) {
        HashMap<Integer,Integer> m = new HashMap<>();
        
        for(int i=0;i<arr.length;i++){
            int rem = (arr[i]%k + k)%k;
            m.put(rem,m.getOrDefault(rem,0)+1);
        }
        
        if(m.getOrDefault(0, 0) % 2 != 0) {
            return false;
        }
        
        for (int i = 1; i <= k / 2; i++) {
            if (!m.getOrDefault(i, 0).equals(m.getOrDefault(k - i, 0))) {
                return false;
            }
        }
        return true;
    }
}